rfcs/0006-transaction.md

```diff
@@ -0,0 +1,299 @@
+
+1. Track the sequence number when transaction started.
+2. Track both read and write keys in the transaction state.
+3. When `commit()` is called, check conflicts with the recent committed transactions in the Oracle which committed after the transaction started.
+4. The conflict check verifies that: the transaction's read and write keys do not conflict with other transactions' write keys
+
+This approach provides Serializable isolation level by detecting both Write-Write conflicts and Read-Write conflicts. It does not rely on MemTable size like RocksDB does, and provides more accurate conflict detection, without worrying about the MemTable size.
+
+The downside is that it requires more memory to track the read keys and recent committed transactions. However, Badger mitigates this by:
+
+1. Using u64 key fingerprints instead of full keys to track read/write keys.
+2. Having an GC mechanism to clean up old committed transactions from the Oracle.
+
+The GC is also straightforward - when all running transactions have sequence numbers greater than a committed transaction's sequence number, that committed transaction can be removed from the Oracle.
+
+## Proposal
+
+### API
+
+The transaction API should be resembles with the non-transactioned API. All the api like `get()`, `put()`, `range()` should also be supported in the transaction API, and these APIs should be aligned with the Snapshot API + WriteBatch API. The only difference is that the transaction API should be able to `commit()` or `rollback()` the changes.
+
+(note: at the time of this RFC, the Snapshot API and WriteBatch API are still not implemented, the transaction feature should be implemented after these APIs are implemented.)
+
+The write operations like `put()` in the transaction should be buffered in the transaction state, and only be written to the underlying storage when `commit()` is called.
+
+```rust
+let object_store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
+let options = DbOptions::default();
+let kv_store = Db::open_with_opts(
+    Path::from("/tmp/test_kv_store"),
+    options,
+    object_store,
+)
+.await
+.unwrap();
+
+// start transaction and write some data
+let txn = kv_store.transaction();
+let val1 = txn.get("key1").await;
+
+// could batch the write operations
+txn.put("key2", val1);
+txn.put("key3", "val3");
+
+// should see the new value of key2
+assert_eq!(txn.get("key2").await, val1);
+
+// commit, might fail on conflict, when it fails, the transaction should be rolled back
+txn.commit().await?;
+```
+
+### Conflict Checking
+
+The approach we choose on conflict checking is expected to be more similar to Badger's global `Oracle` approach, since we hope to provide SSI in the transaction feature, and it's less prone to unnecessarily abort the transaction with limited MemTable size.
+
+Below is the data structure of the global `Oracle` and the transaction state:
+
+```rust
+struct Oracle {
+    next_seq: AtomicU64,
```

Comment:
One more detail that I'm a bit unclear on - how is this value incremented, and how is the start sequence number assigned to read transactions? Also don't we need to differentiate transactions that ask to read committed vs uncommitted data? The way I'm thinking it would work is this:

1. Oracle will either accept or reject writes as part of the queue/channel of writes on the commit pipeline
2. When the Oracle accepts a write, it will assign it `next_seq` and increment `next_seq`. However it can't use `next_seq` - 1 as the `started_seq` for new transactions yet, as the writes haven't been applied to db yet.
3. The commit pipeline applies the write to the db
4. At this point, the commit pipeline needs to notify the oracle that it has applied the write so it can bump the seq num it assigned to new txns for their `started_seq`. So I think we need to store some other watermark like `latest_applied_seq` that tracks the highest sequence number written to the db.
5. Additionally, reads can request that a read level of `Committed` or `Uncommitted`. We can bump the latest applied seq watermark when data is added to the in-memory table, but we need to track some other watermark like `latest_durable_seq` that keeps track of the highest sequence number that is durably stored. This needs to be bumped whenever a wal or memtable flush happens, depending on if wal is enabled/disabled. Then the oracle should hand out this as the `started_seq` if the user requests read level `Committed`.

This brings me to one other small concern - we currently allow setting the read level for each get. When we include gets/scans in a transaction, we should not allow configuring this - this should be specified when initializing the transaction.

Reply:
> Additionally, reads can request that a read level of Committed or Uncommitted.

The naming feels kind of confusing here 🤔 Is my understanding correct that committed / uncommitted in this context refers to whether WAL has been flushed to s3?

Would it make sense to rename those, so they're not mixed up with txn isolation levels?

Reply:
Yeah, your understanding is correct - in this context uncommitted/committed specifies whether the query should include data that is not yet durably stored or only durably stored data, respectively. I agree that the naming is misleading - maybe we call it volatile/durable? Want to open an issue?

Reply:
> This brings me to one other small concern - we currently allow setting the read level for each get. When we include gets/scans in a transaction, we should not allow configuring this - this should be specified when initializing the transaction.

+1 for this. the isolation/read level should be the same inside a transaction

Reply:
>  Want to open an issue?

https://github.com/slatedb/slatedb/issues/369